brion/v2/morphology.cpp

```diff
@@ -0,0 +1,244 @@
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "morphology.h"
+
+#include <brion/morphology.h>
+#include <brion/circuit.h>
+
+#include <math.h>
+
+namespace brion
+{
+namespace v2
+{
+
+namespace
+{
+    void nullDeleter( const brion::Morphology* ) {}
+}
+
+class Morphology::_Impl
+{
+public:
+    boost::shared_ptr< const brion::Morphology > _morphology;
+
+    Vector4fsPtr _points;
+    Vector2isPtr _sections;
+    SectionTypesPtr _types;
+    Vector2isPtr _apicals;
+
+    bool _isGlobal;
+    Vector3f _somaPosition;
+
+    _Impl( const boost::shared_ptr< const brion::Morphology >& morphology,
+           const MorphologyRepairStage stage )
+        : _morphology( morphology )
+        , _points( morphology->readPoints( stage ))
+        , _sections( morphology->readSections( stage ))
+        , _types( morphology->readSectionTypes( ))
+        , _apicals( morphology->readApicals( ))
+        , _isGlobal( false )
+        , _somaPosition( Vector3f::ZERO ) // the soma is assumed to be
+                                          // centered at the 0, 0, 0
+    {
+    }
+
+    uint32_ts getSectionIDs( const SectionTypes& types ) const
+    {
+        bool bits[] = { false, false, false, false, false };
+        for( SectionTypes::const_iterator i = types.begin();
+             i != types.end(); ++i )
+        {
+            bits[size_t( *i )] = true;
+        }
+
+        uint32_ts result;
+        for( uint32_t i = 0; i != _types->size(); ++i )
+        {
+            const SectionType type = ( *_types )[i];
+            assert( type < SECTION_APICAL_DENDRITE + 1 );
+            if( bits[size_t( type )] )
+                result.push_back( i );
+        }
+        return result;
+    }
+
+    Vector4fs getSectionSamples( const size_t sectionID,
+                                 const floats& points = floats( )) const
+    {
+        // If the section is the soma return directly the soma position.
+        if(( *_types )[sectionID] == SECTION_SOMA )
+        {
+            Vector4fs result;
+            for( size_t i = 0; i != points.size(); ++i )
+                result.push_back( _somaPosition );
+            return result;
+        }
+
+        const size_t start = ( *_sections )[sectionID][0];
+        const size_t end = sectionID == _sections->size() - 1 ?
+            _points->size() : ( *_sections )[sectionID + 1][0];
+
+        Vector4fs result;
+
+        // Dealing with the degenerate case of single point sections
+        assert(end > start);
```

Comment:
in the light of broken morphologies, I would refrain of asserting things when they crash your program. Report clearly that there is a mistake with this morphology.